File: examples/lang-julia/TwoStageBenders.jl
# TwoStageBenders.jl
#
# Author
#   Oscar Dowson
#
# Description
#   A simple example reading a StochOptFormat file and solving it via Benders
#   decomposition.
#
#   The code is intended for pedagogical use. It does not contain robust checks
#   or nice error messages, preferring to throw assertion errors.
#
# Usage
#   julia TwoStageBenders.jl [problem]
#   julia TwoStageBenders.jl ../problems/newsvendor.sof.json
#
# Notes
#   You need to install Julia, and have the following packages installed:
#       Clp, JSON, JSONSchema, JuMP, and SHA.
module TwoStageBenders

import Clp
import JSON
import JSONSchema
import JuMP
import Printf
import SHA

const SCHEMA_FILENAME =
    joinpath(dirname(dirname(@__DIR__)), "sof-latest.schema.json")

const RESULT_SCHEMA_FILENAME =
    joinpath(dirname(dirname(@__DIR__)), "sof_result.schema.json")

struct TwoStageProblem
    sha256::String
    first::JuMP.Model
    second::JuMP.Model
    state_variables::Set{String}
    validation_scenarios::Vector{Dict{String, Any}}
end

"""
    TwoStageProblem(filename::String; validate::Bool = true)

Create a two-stage stochastic program by reading `filename` in StochOptFormat.

If `validate`, validate `filename` against the StochOptFormat schema.
"""
function TwoStageProblem(filename::String; validate::Bool = true)
    sha_256 = open(filename) do io
        bytes2hex(SHA.sha2_256(io))
    end
    data = JSON.parsefile(filename)
    if validate
        _validate(data; schema_filename = SCHEMA_FILENAME)
    end
    @assert(data["version"]["major"] == 0)
    @assert(data["version"]["minor"] == 2)
    first, second = _get_stage_names(data)
    problem = TwoStageProblem(
        sha_256,
        _mathoptformat_to_jump(data, first),
        _mathoptformat_to_jump(data, second),
        Set{String}(),
        get(data, "validation_scenarios", Dict{String, Any}[]),
    )
    for k in keys(problem.first.ext[:state_variables])
        push!(problem.state_variables, k)
    end
    _initialize_first_stage(data, first, problem.first)
    return problem
end

function _validate(data::Dict; schema_filename::String)
    schema = JSONSchema.Schema(JSON.parsefile(schema_filename))
    return JSONSchema.validate(data, schema)
end

function _initialize_first_stage(data::Dict, first::String, sp::JuMP.Model)
    for (name, init) in data["root"]["state_variables"]
        x = JuMP.variable_by_name(sp, sp.ext[:state_variables][name]["in"])
        JuMP.fix(x, init["initial_value"]; force = true)
    end
    JuMP.@variable(sp, -1e6 <= theta <= 1e6)
    JuMP.set_objective_function(sp, JuMP.objective_function(sp) + theta)
    return
end

function _get_stage_names(data::Dict)
    @assert length(data["nodes"]) == 2
    @assert length(data["root"]["successors"]) == 1
    first_node, probability = first(data["root"]["successors"])
    @assert probability == 1.0
    successors = data["nodes"][first_node]["successors"]
    @assert length(successors) == 1
    second_node, probability = first(successors)
    @assert probability == 1.0
    @assert length(data["nodes"][second_node]["successors"]) == 0
    return first_node, second_node
end

function _mathoptformat_to_jump(data, name)
    node = data["nodes"][name]
    sp = data["subproblems"][node["subproblem"]]
    model = JuMP.MOI.FileFormats.Model(format = JuMP.MOI.FileFormats.FORMAT_MOF)
    io = IOBuffer()
    write(io, JSON.json(sp["subproblem"]))
    seekstart(io)
    JuMP.MOI.read!(io, model)
    subproblem = JuMP.Model(Clp.Optimizer)
    JuMP.MOI.copy_to(subproblem, model)
    JuMP.set_silent(subproblem)
    subproblem.ext[:state_variables] = sp["state_variables"]
    subproblem.ext[:realizations] = node["realizations"]
    _convert_realizations(subproblem.ext[:realizations])
    return subproblem
end

# These `_convert_realization` functions are Julia-specific helper functions to
# convert the support dictionaries from `Dict{Symbol, Any}` to
# `Dict{Symbol, Float64}`.
function _convert_realizations(realizations)
    for r in realizations
        r["support"] = convert(Dict{String, Float64}, r["support"])
    end
end

# These are helper functions to extract the incoming and outgoing state
# variables from a JuMP model.
function _incoming_state(sp::JuMP.Model, name::String)
    return JuMP.variable_by_name(sp, sp.ext[:state_variables][name]["in"])
end
function _outgoing_state(sp::JuMP.Model, name::String)
    return JuMP.variable_by_name(sp, sp.ext[:state_variables][name]["out"])
end

function _solve_first_stage(problem::TwoStageProblem)
    sp = problem.first
    JuMP.optimize!(sp)
    @assert JuMP.termination_status(sp) == JuMP.MOI.OPTIMAL
    return Dict(
        "objective" => JuMP.objective_value(sp) - JuMP.value(sp[:theta]),
        "primal" => Dict(
            JuMP.name(x) => JuMP.value(x) for x in JuMP.all_variables(sp)
            # Don't leak the cost-to-go variable back to the user.
            if x != sp[:theta]
        )
    )
end

function _solve_second_stage(
    problem::TwoStageProblem,
    state_variables::Dict{String, Float64},
    random_variables::Dict{String, Float64},
)
    sp = problem.second
    for name in problem.state_variables
        JuMP.fix(_incoming_state(sp, name), state_variables[name]; force = true)
    end
    for (name, w) in random_variables
        JuMP.fix(JuMP.variable_by_name(sp, name), w; force = true)
    end
    JuMP.optimize!(sp)
    return Dict(
        "objective" => JuMP.objective_value(sp),
        "primal" => Dict(
            JuMP.name(x) => JuMP.value(x) for x in JuMP.all_variables(sp)),
    )
end

function _add_benders_optimality_cut(
    problem::TwoStageProblem,
    state_variables::Dict{String, Float64},
    probabilities::Vector{Float64},
    objectives::Vector{Float64},
    dual_variables::Vector{Dict{String, Float64}},
)
    cut_term = JuMP.@expression(problem.first, sum(
            p * (
                y + sum(
                    π[name] * (_outgoing_state(problem.first, name) - x_val)
                    for (name, x_val) in state_variables
                )
            ) for (π, y, p) in zip(dual_variables, objectives, probabilities)
        )
    )
    if JuMP.objective_sense(problem.first) == JuMP.MOI.MAX_SENSE
        JuMP.@constraint(problem.first, problem.first[:theta] <= cut_term)
    else
        JuMP.@constraint(problem.first, problem.first[:theta] >= cut_term)
    end
    return
end

"""
    train(problem::TwoStageProblem; iteration_limit = 20)

Train a two-stage problem using Benders decomposition.
"""
function train(problem::TwoStageProblem; iteration_limit = 20)
    println("Iteration | Lower Bound | Upper Bound | Gap (abs)")
    for iter = 1:iteration_limit
        ret_first = _solve_first_stage(problem)
        probabilities = Float64[]
        objectives = Float64[]
        dual_variables = Dict{String, Float64}[]
        x = Dict(
            name => ret_first["primal"][s["out"]]
            for (name, s) in problem.second.ext[:state_variables]
        )
        for realization in problem.second.ext[:realizations]
            ret = _solve_second_stage(problem, x, realization["support"])
            push!(probabilities, realization["probability"])
            push!(objectives, ret["objective"])
            push!(dual_variables, Dict(
                name => JuMP.shadow_price(
                    JuMP.FixRef(_incoming_state(problem.second, name))
                ) for name in problem.state_variables
            ))
        end
        deterministic_bound = JuMP.objective_value(problem.first)
        stat_bound = ret_first["objective"] +
            sum(p * o for (p, o) in zip(probabilities, objectives))
        is_min = JuMP.objective_sense(problem.first) == JuMP.MOI.MIN_SENSE
        gap = abs(deterministic_bound - stat_bound)
        Printf.@printf(
            "%9d | % 5.4e | % 5.4e | %4.3e\n",
            iter,
            is_min ? deterministic_bound : stat_bound,
            is_min ? stat_bound : deterministic_bound,
            gap,
        )
        if gap < 1e-6
            println("Terminating training: convergence")
            return
        end
        _add_benders_optimality_cut(
            problem, x, probabilities, objectives, dual_variables
        )
    end
    println("Terminating training: iteration limit")
    return
end

"""
    evaluate(
        problem::TwoStageProblem;
        scenarios = problem.validation_scenarios,
    )

Evaluate the policy after training `problem` on `scenarios`. By default, these
are the `validation_scenarios` contained in the StochOptFormat file, but you can
pass a different set if necessary.
"""
function evaluate(
    problem::TwoStageProblem;
    scenarios = problem.validation_scenarios,
    filename::Union{Nothing, String} = nothing
)
    solutions = Vector{Dict{String, Any}}[]
    for s_dict in scenarios
        scenario = s_dict["scenario"]
        @assert length(scenario) == 2
        first_sol = _solve_first_stage(problem)
        incoming_state = Dict(
            name => first_sol["primal"][s["out"]]
            for (name, s) in problem.second.ext[:state_variables]
        )
        second_sol = _solve_second_stage(
            problem,
            incoming_state,
            convert(Dict{String, Float64}, scenario[2]["support"]),
        )
        push!(solutions, [first_sol, second_sol])
    end
    solution = Dict(
        "problem_sha256_checksum" => problem.sha256,
        "scenarios" => solutions,
    )
    _validate(solution; schema_filename = RESULT_SCHEMA_FILENAME)
    if filename !== nothing
        open(filename, "w") do io
            write(io, JSON.json(solution))
        end
    end
    return solution
end

export
    TwoStageProblem,
    train,
    evaluate

end # module TwoStageBenders

if endswith(@__FILE__, PROGRAM_FILE)
    TSSP = TwoStageBenders
    @assert length(ARGS) == 1
    filename = ARGS[1]
    problem = TSSP.TwoStageProblem(filename)
    ret = TSSP.train(problem; iteration_limit = 20)
    solutions = TSSP.evaluate(problem; filename = "sol_jl.json")
    if endswith(filename, "news_vendor.sof.json")
        # Check solutions
        @assert solutions["scenarios"][1][1]["objective"] ≈ -10
        @assert solutions["scenarios"][1][2]["objective"] ≈ 15
        @assert solutions["scenarios"][2][2]["objective"] ≈ 15
        @assert solutions["scenarios"][3][2]["objective"] ≈ 13.5
    end
end

File: examples/lang-python/TwoStageBenders.py
# Copyright (c) 2020: Oscar Dowson, Joaquim Dias Garcia, and contributors
#
# Use of this source code is governed by an MIT-style license that can be found
# in the LICENSE.md file or at https://opensource.org/licenses/MIT.

# TwoStageBenders.py
#
# Author
#   Oscar Dowson
#
# Description
#   A simple example reading a StochOptFormat file and solving it via Benders
#   decomposition.
#
#   The code is intended for pedagogical use. It does not contain robust checks
#   or nice error messages, preferring to throw assertion errors.
#
# Usage
#   python TwoStageBenders.py [problem]
#   python TwoStageBenders.py ../problems/news_vendor.sof.json
#
# Notes
#   You need to install python, and have the following packages installed:
#       hashlib, jsonschema, and pulp.

import hashlib
import json
import jsonschema
import math
import os
import pulp

class TwoStageProblem:
    def __init__(self, filename, validate = True):
        with open(filename, 'rb') as io:
            self.sha256 = hashlib.sha256(io.read()).hexdigest()
        with open(filename, 'r') as io:
            self.data = json.load(io)
        version_dir = os.path.join(
            os.path.dirname(
                os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
            ),
            'schemas',
        )
        self.schema = os.path.join(version_dir, 'sof-1.schema.json')
        self.result_schema = os.path.join(version_dir, 'sof-result.schema.json')
        if validate:
            self._validate_stochoptformat()
        first, second = self._get_stage_names()
        self.first = self._mathoptformat_to_pulp(first)
        self.second = self._mathoptformat_to_pulp(second)
        self._initialize_first_stage(first)
        return

    def train(self, iteration_limit):
        print('Iteration | Lower Bound | Upper Bound | Gap (abs)')
        for iter in range(iteration_limit):
            ret_first = self._solve_first_stage()
            probabilities = []
            objectives = []
            dual_variables = []
            x = {
                name: ret_first['primal'][s['out']]
                for (name, s) in self.second['state_variables'].items()
            }
            for realization in self.second.get('realizations', []):
                ret = self._solve_second_stage(x, realization['support'])
                probabilities.append(realization['probability'])
                objectives.append(ret['objective'])
                dual_variables.append({
                    name: self._incoming_state(self.second, name).dj
                    for name in self.second['state_variables']
                })
            deterministic_bound = pulp.value(self.first['subproblem'].objective)
            stat_bound = ret_first['objective'] + sum(
                p * o for (p, o) in zip(probabilities, objectives)
            )
            is_min = self.first['subproblem'].sense == 1
            gap = abs(deterministic_bound - stat_bound)
            print(
                '%9d | % 5.4e | % 5.4e | %4.3e' % (
                    iter + 1,
                    deterministic_bound if is_min else stat_bound,
                    stat_bound if is_min else deterministic_bound,
                    gap
                )
            )
            if abs(deterministic_bound - stat_bound) < 1e-6:
                print('Terminating training: convergence')
                return
            self._add_benders_optimality_cut(
                x, probabilities, objectives, dual_variables
            )
        print('Terminating training: iteration limit')
        return

    def evaluate(self, scenarios = None, filename = None):
        if scenarios is None:
            scenarios = self.data['validation_scenarios']
        solutions = []
        for scenario in scenarios:
            assert len(scenario) == 2
            first_sol = self._solve_first_stage()
            incoming_state = {
                name: first_sol['primal'][s['out']]
                for (name, s) in self.second['state_variables'].items()
            }
            second_sol = self._solve_second_stage(
                incoming_state, scenario[1].get('support', {})
            )
            solutions.append([first_sol, second_sol])
        solution = {
            'problem_sha256_checksum': self.sha256,
            'scenarios': solutions
        }
        with open(self.result_schema, 'r') as io:
            schema = json.load(io)
        jsonschema.validate(instance = solution, schema = schema)
        if filename is not None:
            with open(filename, 'w') as io:
                json.dump(solution, io)
        return solution

    def _validate_stochoptformat(self):
        with open(self.schema, 'r') as io:
            schema = json.load(io)
        return jsonschema.validate(instance = self.data, schema = schema)

    def _get_stage_names(self):
        assert len(self.data['nodes']) == 2
        assert len(self.data['root']['successors']) == 1
        first_node, probability = \
            next(iter(self.data['root']['successors'].items()))
        assert probability == 1.0
        successors = self.data['nodes'][first_node]['successors']
        assert len(successors) == 1
        second_node, probability = next(iter(successors.items()))
        assert probability == 1.0
        assert len(self.data['nodes'][second_node].get('successors', [])) == 0
        return first_node, second_node

    def _mathoptformat_to_pulp(self, name):
        node = self.data['nodes'][name]
        subproblem = self.data['subproblems'][node['subproblem']]
        sp = subproblem['subproblem']
        # Create the problem
        if sp['objective']['sense'] == 'max':
            prob = pulp.LpProblem(name, pulp.LpMaximize)
        else:
            prob = pulp.LpProblem(name, pulp.LpMinimize)        
        # Initialize the variables
        vars = {x['name']: pulp.LpVariable(x['name']) for x in sp['variables']}
        # Add the objective function
        obj = sp['objective']['function']
        if obj['type'] == 'Variable':
            prob += vars[obj['name']]
        elif obj['type'] == 'ScalarAffineFunction':
            prob += obj['constant'] + pulp.lpSum(
                term['coefficient'] * vars[term['variable']] for 
                term in obj['terms']
            )
        else:
            raise(Exception('Unsupported objective: ' + str(obj)))
        # Add the constraints
        for c in sp['constraints']:
            f, s = c['function'], c['set']
            if f['type'] == 'Variable':
                x = f['name']
                if s['type'] == 'GreaterThan':
                    vars[x].lowBound = s['lower']
                elif s['type'] == 'LessThan':
                    vars[x].upBound = s['upper']
                elif s['type'] == 'EqualTo':
                    vars[x].lowBound = s['value']
                    vars[x].upBound = s['value']
                elif s['type'] == 'Interval':
                    vars[x].lowBound = s['lower']
                    vars[x].upBound = s['upper']
                else:
                    raise(Exception('Unsupported set: ' + str(s)))
            elif f['type'] == 'ScalarAffineFunction':
                lhs = f['constant'] + pulp.lpSum(
                    term['coefficient'] * vars[term['variable']] for
                    term in f['terms']
                )
                if s['type'] == 'GreaterThan':
                    prob += lhs >= s['lower']
                elif s['type'] == 'LessThan':
                    prob += lhs <= s['upper']
                elif s['type'] == 'EqualTo':
                    prob += lhs == s['value']
                elif s['type'] == 'Interval':
                    prob += lhs <= s['upper']
                    prob += lhs >= s['lower']
                else:
                    raise(Exception('Unsupported set: ' + str(s)))
            else:
                raise(Exception('Unsupported function: ' + str(f)))
        return {
            'subproblem': prob,
            'vars': vars,
            'state_variables': subproblem['state_variables'],
            'realizations': node.get('realizations', []),
        }

    def _incoming_state(self, sp, name):
        return sp['vars'][sp['state_variables'][name]['in']]

    def _outgoing_state(self, sp, name):
        return sp['vars'][sp['state_variables'][name]['out']]

    def _initialize_first_stage(self, name):
        for (name, init) in self.data['root']['state_variables'].items():
            x = self.first['vars'][self.first['state_variables'][name]['in']]
            x.lowBound = init
            x.upBound = init
        self.first['theta'] = pulp.LpVariable('theta', -10**6, 10**6)
        self.first['subproblem'].objective += self.first['theta']
        return

    def _solve_first_stage(self):
        self.first['subproblem'].solve(pulp.coin_api.PULP_CBC_CMD(msg = 0))
        return {
            'objective': pulp.value(self.first['subproblem'].objective) -
                         self.first['theta'].varValue,
            'primal': {
                v.name: v.varValue for v in self.first['subproblem'].variables()
            }
        }

    def _solve_second_stage(self, state_variables, random_variables):
        for (name, val) in state_variables.items():
            v = self._incoming_state(self.second, name)
            v.lowBound = state_variables[name]
            v.upBound = state_variables[name]
        for (name, w) in random_variables.items():
            p = self.second['vars'][name]
            p.lowBound = w
            p.upBound = w
        solver = pulp.coin_api.PULP_CBC_CMD(msg = 0)
        self.second['subproblem'].solve(solver)
        return {
            'objective': pulp.value(self.second['subproblem'].objective),
            'primal': {
                v.name: v.varValue for v in self.first['subproblem'].variables()
            }
        }

    def _add_benders_optimality_cut(
        self, state_variables, probabilities, objectives, dual_variables
    ):
        cut_term = pulp.lpSum(
            p * o + p * pulp.lpSum(
                d[name] * (self._outgoing_state(self.first, name) - x_val)
                for (name, x_val) in state_variables.items()
            ) for (p, o, d) in zip(probabilities, objectives, dual_variables)
        )
        if self.first['subproblem'].sense == -1:
            self.first['subproblem'] += self.first['theta'] <= cut_term
        else:
            self.first['subproblem'] += self.first['theta'] >= cut_term
        return

if __name__ == '__main__':
    import sys
    assert len(sys.argv) == 2 
    filename = sys.argv[1]
    problem = TwoStageProblem(filename)
    problem.train(iteration_limit = 20)
    solutions = problem.evaluate(filename = 'sol_py.json')
    if filename.endswith('news_vendor.sof.json'):
        assert solutions['scenarios'][0][0]['objective'] == -10
        assert solutions['scenarios'][0][1]['objective'] == 15
        assert solutions['scenarios'][1][1]['objective'] == 15
        assert solutions['scenarios'][2][1]['objective'] == 13.5

